### PR TITLE
update tracing and logging documentation

### PR DESCRIPTION
**Pull request by RubBra**

### CODE REVIEW

File: doc/logging.md [hunk 1]
```diff
@@ -1,27 +1,23 @@
 # VCML Logging
-VCML uses a centralized logging system that allows its individual components
+VCML uses a centralized logging system provided by MWR that allows its individual components
```
Comment (aut0) [inline doc/logging.md:2]: "provided by mwr" or "libmwr"? I would add a link to the mwr github.

File: doc/logging.md [hunk 2]
```diff
@@ -30,125 +26,108 @@ A full log message could therefore look like this:
 This tells us that there was a debug message at *1s + 200ns*, informing us that
 the component `system.uart0` has been disabled.
 
-Log messages can be generated using the global top level logging functions from
-the `vcml` namespace. These functions use the current SystemC process name as
+Log messages can be generated using the global top level logging function-like macros from
+the `mwr` namespace. These macros use the file and line name as
```
Comment (aut0) [inline doc/logging.md:30]: file name and line number?

File: doc/logging.md [hunk 3]
```diff
@@ -30,125 +26,108 @@ A full log message could therefore look like this:
...
-    };
-}
 ```
 
 ----
 ## Logging Output
-In order to actually see log messages, you need a logger that will receive the
-messages and display them in some form. VCML currently provides three loggers:
+In order to actually see log messages, you need a publisher backend that will receive the
+messages and display them in some form. MWR currently provides three backends:
```
Comment (aut0) [inline doc/logging.md:46]: Für mich ist MWR = MachineWare.
Vllt libmwr? @janweinstock ?

File: doc/logging.md [hunk 4]
```diff
@@ -30,125 +26,108 @@ A full log message could therefore look like this:
...
-| `print_backtrace`   | Print backtraces for logged reports         | `true`  |
+| Print control       | Description                               | Default |
+|---------------------|-------------------------------------------|---------|
+| `print_timestamp`   | Print the timestamp of the log message    | `true`  |
+| `print_sender`      | Print sending process or module           | `true`  |
+| `print_source`      | Print the source location of the log call | `false` |
+| `print_backtrace`   | Print backtraces for logged reports       | `true`  |
 
-It is also possible to create your own logger if you need to process VCML log
+It is also possible to create your own publisher if you need to process MWR log
```
Comment (aut0) [inline doc/logging.md:91]: MWR

File: doc/logging.md [hunk 5]
```diff
@@ -30,125 +26,108 @@ A full log message could therefore look like this:
...
+* `--log-stdout`: Send log output to stdout
+* `--log-file`: Send log output to file
 * `--log-debug`: elevates the log level of all loggers to `LOG_DEBUG`
-* `--log-delta`: toggles `vcml::logger::print_delta_cycle`
 
 ----
 ## Exceptions
-The logging system is typically also used for exception reporting. VCML
-generally uses the class `vcml::report` for exceptions, so a basic approach to
+The logging system is typically also used for exception reporting. MWR
```
Comment (aut0) [inline doc/logging.md:115]: MWR

File: doc/tracing.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+VCML currently has two tracing output options:
+
+* `--trace-stdout`: Send tracing output to stdout.
+* `-t <file>` or `--trace <file>`: send tracing output to file or stderr if no `<file>` is specified.
+
+### Tracing Message Format
+
+`[<PROTOCOL> <TIME>] <SOCKET> <DIR> <OP> <ADDRESS> [<VALUE>] (<RESPONSE>)`
+
+Here, `<PROTOCOL>` refers to the specific protocol that was used, `<TIME>` is a transaction timestamp,
```
Comment (aut0) [inline doc/tracing.md:20]: I prefer one sentence per line or newline at ,
Easier to diff later.

File: doc/logging.md [hunk 7]
```diff
@@ -157,37 +136,10 @@ int sc_main(int argc, char** argv) {
...
-the internal transaction buffer. `<RESPONSE>` holds the current state of the
-transaction, see the TLM reference manual for details. Following is an example
-of an actual TLM transaction trace:
-
-* reading four bytes from address 0x100: `<< RD 0x0000100 [00 FF 00 FF] (TLM_OK_RESPONSE)`
-* writing single byte to invalid address: `<< WR 0xFFFFFFFF [EE] (TLM_ADDRESS_ERROR_RESPONSE)`
+`mwr::publisher::print_backtrace = false`.
 
 ----
 Documentation April 2020
```
Comment (aut0) [inline doc/logging.md:145]: This should be updated

Comment (aut0) [reply doc/logging.md:145]: Also in main.md

**Review by aut0 — CHANGES_REQUESTED** [review]

File: doc/logging.md [hunk 8]
```diff
@@ -30,125 +26,108 @@ A full log message could therefore look like this:
...
 
-* `vcml::log_term`: write the log message to `stderr` using color if available.
-* `vcml::log_file`: write log messages to a file specified during construction.
-* `vcml::log_stream`: sends the log messages to any `std::ostream` instance.
+* `mwr::publishers::terminal`: write the log message to `stderr` using color if available.
+* `mwr::publishers::file`: write log messages to a file specified during construction.
+* `mwr::publishers::stream`: sends the log messages to any `std::ostream` instance.
 
-Each logger can furthermore also specify the log levels it is interested in.
+Each backend can furthermore also specify the log levels it is interested in.
```
Comment (janweinstock) [inline doc/logging.md:52]: Each publisher?

File: doc/logging.md [hunk 9]
```diff
@@ -30,125 +26,108 @@ A full log message could therefore look like this:
...
 ```
-vcml::log_file my_logger("log.txt");
-my_logger.set_level(vcml::LOG_ERROR); // only receives error messages
-my_logger.set_level(vcml::LOG_ERROR, vcml::LOG_INFO); // receives error, warning and info messages
+mwr::publishers::file my_logger("log.txt");
+my_logger.set_level(mwr::LOG_ERROR); // only receives error messages
+my_logger.set_level(mwr::LOG_ERROR, mwr::LOG_INFO); // receives error, warning and info messages
 ```
 
 Having multiple loggers for different purposes is possible and encouraged. You
```
Comment (janweinstock) [inline doc/logging.md:61]: multiple publishers

File: doc/tracing.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+`<SOCKET>` the name of the specific socket involved in the transaction, `<DIR>` refers to the direction of the transaction. Outgoing requests are
+denoted with a `>>`, incoming responses use `<<`. `<OP>` refers to the access
+operation and can be `RD` (read) or `WR` (write). `<ADDRESS>` shows the
+destination address in hex and `<VALUE>` contains the individual byte values of
+the internal transaction buffer. `<RESPONSE>` holds the current state of the
+transaction, see the TLM reference manual for details. Following is an example
+of an actual TLM transaction trace:
+
+* reading four bytes from address 0x100: `[TLM 10.910098127] system.uart0.thr << RD 0x0000100 [00 FF 00 FF] (TLM_OK_RESPONSE)`
+* writing single byte to invalid address: `[TLM 10.910098127] system.cpu.hart0.data << WR 0xFFFFFFFF [EE] (TLM_ADDRESS_ERROR_RESPONSE)`
```
Comment (janweinstock) [inline doc/tracing.md:30]: Should use lower-case hex numbers for consistency

File: doc/tracing.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+Tracing is available for all known payload types: `tlm_generic_payload`, `gpio_payload`,
+`clk_payload`, `pci_payload`, `i2c_payload`, `spi_payload`, `sd_command`, `sd_data`,
+`vq_message`, `serial_payload`, `eth_frame`, `can_frame`
+
+## Tracing Output
+VCML currently has two tracing output options:
+
+* `--trace-stdout`: Send tracing output to stdout.
+* `-t <file>` or `--trace <file>`: send tracing output to file or stderr if no `<file>` is specified.
```
Comment (janweinstock) [inline doc/tracing.md:14]: Must specify <file>, remove "or stderr if no..."

File: doc/tracing.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
+# VCML Tracing
+
+VCML offers TLM transaction tracing functionality.
```
Comment (janweinstock) [inline doc/tracing.md:3]: VCML offers transaction tracing functionality for TLM and various other protocols, such as GPIO, CLK, PCI, I2C, SPI, SD, Serial, Ethernet, CAN and Virtio.

**Review by janweinstock — CHANGES_REQUESTED** [review]

File: doc/logging.md (hunk 8 above)
Comment (RubBra) [reply doc/logging.md:52]: auch bei logging, da heißen die ja nicht explizit so?

File: doc/logging.md (hunk 1 above)
Comment (aut0) [reply doc/logging.md]: Please add a link to the mwr github repo for the word "libmwr"

File: doc/logging.md [hunk 13]
```diff
@@ -1,154 +1,134 @@
 # VCML Logging
-VCML uses a centralized logging system that allows its individual components
+VCML uses a centralized logging system provided by libmwr that allows its individual components
```
Comment (aut0) [inline doc/logging.md:2]: Add link to mwr GitHub on libmwr

**Review by aut0 — CHANGES_REQUESTED** [review]

File: doc/tracing.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
+# VCML Tracing
+
+VCML offers transaction tracing functionality for TLM and various other protocols, 
+such as GPIO, CLK, PCI, I2C, SPI, SD, Serial, Ethernet, CAN and Virtio.
+Tracing can be enabled globally via `system.trace=1` or for individual components via e.g. `system.sdhci.sd_out.trace=1`
```
Comment (aut0) [inline doc/tracing.md:5]: .

File: doc/tracing.md [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+Tracing is available for all known payload types: `tlm_generic_payload`, `gpio_payload`,
+`clk_payload`, `pci_payload`, `i2c_payload`, `spi_payload`, `sd_command`, `sd_data`,
+`vq_message`, `serial_payload`, `eth_frame`, `can_frame`
```
Comment (aut0) [inline doc/tracing.md:9]: .

File: doc/logging.md (hunk 2 above)
Comment (RubBra) [reply doc/logging.md]: it doesnt, the macro confused me. corrected

File: doc/logging.md (hunk 8 above)
Comment (janweinstock) [reply doc/logging.md:52]: Ich meinte die "backends" beim logging heißen "publisher", deswegen würde ich "Each publisher can furthermore..." schreiben.

**Review by janweinstock — CHANGES_REQUESTED** [review]

**Review by janweinstock — APPROVED** [review]

**Review by aut0 — APPROVED** [review]

**Pull request merged**